webui/src/lib/hooks/router.ts

```diff
@@ -0,0 +1,52 @@
+import {
+  useHistory,
+  useLocation,
+  useParams,
+  generatePath,
+} from "react-router-dom";
+
+export const useQuery = <T extends Partial<T>>(): T => {
+  const location = useLocation();
+  const result: Record<string, string> = {};
+  for (const [key, value] of new URLSearchParams(location.search).entries()) {
+    result[key] = value;
+  }
+  return result as T;
```

Comment:
I don't remember much JavaScript, but isn't this just
```suggestion
  return Object.fromEntries(new URLSearchParams(location.search).entries()) at T;
```
?

Also not sure about `T extends Partial<T>`; is this a way of restricting the requested type to be an object?

Also note that both versions continue the usual TypeScript lie about object keys.  They are not really strings, they can be numbers as well.  For instance:
```
>>> Object.fromEntries(new URLSearchParams("c=d&e=f&12=13").entries())
Object { 12: "13", "c": "d", e: "f" }
```
has a _number_ key 12, and depends on an implicit conversion to make it seem like it has a string key. 🤷🏽

Reply:
It does work, but the [URL spec](https://url.spec.whatwg.org/#interface-urlsearchparams) defines keys and values as strings. I'm guessing the automated conversion is done for convenience, but that's implementation-specific and doesn't _have_ to be the case in all browsers, even if currently it is 😊

Reply:
Yeah, I dug around the ECMAScript standard and JavaScript objects were standardized as a mess.  A JavaScript object can have keys of types symbol, number, or string.  Symbols are symbols because of their types, but a key is a number because if it is convertible to a number 🤦🏽 .  Welcome to JavaScript.

[#not-my-meme
![welcome to JavaScript where the rules don't matter and datatypes are made up](https://user-images.githubusercontent.com/7802932/201018470-67de3613-530a-401d-8ba1-48da5ee0ac89.png)](https://www.memecreator.org/meme/welcome-to-javascript-where-rules-dont-matter-and-datatypes-are-made-up/)

StilLGTM, of course!